GraphQL random CDI context issues when using RestClients with @Source

### Describe the bug

With version 2.10.x when using @Source and multiple RestClients there is a problem when accessing @RequestScoped elements like RoutingContext.

The issue is random and does not always break, so for that i have made a reproducer project with an integration test that you can check out.
https://github.com/yuxblank/quarkus-requestscope-bug


The error it's sometime different (maybe secondary cause) but mostly:

```log
uppressed: javax.ws.rs.ProcessingException: javax.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()
		at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:780)
		at org.jboss.resteasy.microprofile.client.impl.MpClientInvocation.filterRequest(MpClientInvocation.java:75)
		at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:491)
		at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:152)
		at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
		at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
		at com.sun.proxy.$Proxy76.getExample(Unknown Source)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
		at org.jboss.resteasy.microprofile.client.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:146)
		at com.sun.proxy.$Proxy77.getExample(Unknown Source)
		at com.rgigroup.quarkuscve.GraphQLEndpoint.exampleSource(GraphQLEndpoint.kt:24)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:566)
		at io.smallrye.graphql.execution.datafetcher.helper.ReflectionInvoker.invoke(ReflectionInvoker.java:92)
		at io.quarkus.smallrye.graphql.runtime.spi.datafetcher.QuarkusDefaultDataFetcher.lambda$invokeAndTransformBlocking$0(QuarkusDefaultDataFetcher.java:62)
		at io.smallrye.context.impl.wrappers.SlowContextualCallable.call(SlowContextualCallable.java:21)
		at io.quarkus.smallrye.graphql.runtime.spi.datafetcher.BlockingHelper.lambda$runBlocking$0(BlockingHelper.java:27)
		at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
		at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
		at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
		at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
		at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:555)
		at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
		at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
		at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
		at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
		at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
		at java.base/java.lang.Thread.run(Thread.java:834)
	Caused by: javax.enterprise.inject.IllegalProductException: Normal scoped producer method may not return null: io.quarkus.vertx.http.runtime.CurrentVertxRequest.getCurrent()
		at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
		at io.quarkus.vertx.http.runtime.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_Bean.create(Unknown Source)
		at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:73)
		at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:28)
		at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.arc$delegate(Unknown Source)
		at io.vertx.ext.web.CurrentVertxRequest_ProducerMethod_getCurrent_6dc23d16d53ba5c34e1e7b6f54290fd7b9aebd76_ClientProxy.request(Unknown Source)
		at com.rgigroup.quarkuscve.JWTAwareContext.parseClaims(JWTAwareContext.kt:29)
		at com.rgigroup.quarkuscve.JWTAwareContext.getTenant(JWTAwareContext.kt:21)
		at com.rgigroup.quarkuscve.JWTAwareContext_ClientProxy.getTenant(Unknown Source)
		at com.rgigroup.quarkuscve.ClientInterceptor.filter(ClientInterceptor.kt:19)
		at com.rgigroup.quarkuscve.ClientInterceptor_ClientProxy.filter(Unknown Source)
		at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:767)
		... 32 more

```





### Expected behavior

Current server request should be accessible during the context of a request.

### Actual behavior

With 2.10.x it randomly fails.
With 2.9.x works flawlessly

### How to Reproduce?

https://github.com/yuxblank/quarkus-requestscope-bug

- run mvn:test
- some run will break
- change quarkus version to 2.9.x
- run mvn:test
- test will never throw

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

adopt-openjdk-11.0.8

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.10.x

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

--- a/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/AbstractAsyncDataFetcher.java
+++ b/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/AbstractAsyncDataFetcher.java
@@ -29,34 +29,40 @@ public abstract class AbstractAsyncDataFetcher<K, T> extends AbstractDataFetcher
             DataFetcherResult.Builder<Object> resultBuilder,
             Object[] transformedArguments) throws Exception {
 
-        Uni<?> uni = handleUserMethodCall(dfe, transformedArguments);
-        return (O) uni
-                .onItemOrFailure()
-                .transformToUni((result, throwable, emitter) -> {
-                    if (throwable != null) {
-                        eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), throwable);
-                        if (throwable instanceof GraphQLException) {
-                            GraphQLException graphQLException = (GraphQLException) throwable;
-                            errorResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
-                        } else if (throwable instanceof Exception) {
-                            emitter.fail(SmallRyeGraphQLServerMessages.msg.dataFetcherException(operation, throwable));
-                            return;
-                        } else if (throwable instanceof Error) {
-                            emitter.fail(throwable);
-                            return;
+        ManagedContext requestContext = Arc.container().requestContext();
+        try {
+            RequestContextHelper.reactivate(requestContext, dfe);
+            Uni<?> uni = handleUserMethodCall(dfe, transformedArguments);
+            return (O) uni
+                    .onItemOrFailure()
+                    .transformToUni((result, throwable, emitter) -> {
+                        if (throwable != null) {
+                            eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), throwable);
+                            if (throwable instanceof GraphQLException) {
+                                GraphQLException graphQLException = (GraphQLException) throwable;
+                                errorResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
+                            } else if (throwable instanceof Exception) {
+                                emitter.fail(SmallRyeGraphQLServerMessages.msg.dataFetcherException(operation, throwable));
+                                return;
+                            } else if (throwable instanceof Error) {
+                                emitter.fail(throwable);
+                                return;
+                            }
+                        } else {
+                            try {
+                                resultBuilder.data(fieldHelper.transformOrAdaptResponse(result, dfe));
+                            } catch (AbstractDataFetcherException te) {
+                                te.appendDataFetcherResult(resultBuilder, dfe);
+                            }
                         }
-                    } else {
-                        try {
-                            resultBuilder.data(fieldHelper.transformOrAdaptResponse(result, dfe));
-                        } catch (AbstractDataFetcherException te) {
-                            te.appendDataFetcherResult(resultBuilder, dfe);
-                        }
-                    }
 
-                    emitter.complete(resultBuilder.build());
-                })
-                .subscribe()
-                .asCompletionStage();
+                        emitter.complete(resultBuilder.build());
+                    })
+                    .subscribe()
+                    .asCompletionStage();
+        } finally {
+            requestContext.deactivate();
+        }
     }
 
     protected abstract Uni<?> handleUserMethodCall(DataFetchingEnvironment dfe, final Object[] transformedArguments)
@@ -76,7 +82,7 @@ public abstract class AbstractAsyncDataFetcher<K, T> extends AbstractDataFetcher
     protected CompletionStage<List<T>> invokeBatch(DataFetchingEnvironment dfe, Object[] arguments) {
         ManagedContext requestContext = Arc.container().requestContext();
         try {
-            BlockingHelper.reactivate(requestContext, dfe);
+            RequestContextHelper.reactivate(requestContext, dfe);
             return handleUserBatchLoad(dfe, arguments)
                     .subscribe().asCompletionStage();
         } catch (Exception ex) {

--- a/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/BlockingHelper.java
+++ b/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/BlockingHelper.java
@@ -2,15 +2,15 @@ package io.quarkus.smallrye.graphql.runtime.spi.datafetcher;
 
 import java.util.concurrent.Callable;
 
-import graphql.schema.DataFetchingEnvironment;
-import io.quarkus.arc.InjectableContext;
-import io.quarkus.arc.ManagedContext;
 import io.smallrye.graphql.schema.model.Execute;
 import io.smallrye.graphql.schema.model.Operation;
 import io.vertx.core.Context;
 import io.vertx.core.Promise;
 
-public class BlockingHelper {
+public final class BlockingHelper {
+
+    private BlockingHelper() {
+    }
 
     public static boolean blockingShouldExecuteNonBlocking(Operation operation, Context vc) {
         // Rule is that by default this should execute blocking except if marked as non-blocking)
@@ -32,16 +32,5 @@ public class BlockingHelper {
                 future.fail(ex);
             }
         }, result);
-    }
-
-    public static void reactivate(ManagedContext requestContext, DataFetchingEnvironment dfe) {
-        if (!requestContext.isActive()) {
-            Object maybeState = dfe.getGraphQlContext().getOrDefault("state", null);
-            if (maybeState != null) {
-                requestContext.activate((InjectableContext.ContextState) maybeState);
-            } else {
-                requestContext.activate();
-            }
-        }
     }
 }

--- a/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/QuarkusDefaultDataFetcher.java
+++ b/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/QuarkusDefaultDataFetcher.java
@@ -30,22 +30,34 @@ public class QuarkusDefaultDataFetcher<K, T> extends DefaultDataFetcher<K, T> {
     public <T> T invokeAndTransform(DataFetchingEnvironment dfe, DataFetcherResult.Builder<Object> resultBuilder,
             Object[] transformedArguments) throws Exception {
 
-        Context vc = Vertx.currentContext();
-        if (runBlocking(dfe) || BlockingHelper.blockingShouldExecuteNonBlocking(operation, vc)) {
-            return super.invokeAndTransform(dfe, resultBuilder, transformedArguments);
-        } else {
-            return invokeAndTransformBlocking(dfe, resultBuilder, transformedArguments, vc);
+        ManagedContext requestContext = Arc.container().requestContext();
+        try {
+            RequestContextHelper.reactivate(requestContext, dfe);
+            Context vc = Vertx.currentContext();
+            if (runBlocking(dfe) || BlockingHelper.blockingShouldExecuteNonBlocking(operation, vc)) {
+                return super.invokeAndTransform(dfe, resultBuilder, transformedArguments);
+            } else {
+                return invokeAndTransformBlocking(dfe, resultBuilder, transformedArguments, vc);
+            }
+        } finally {
+            requestContext.deactivate();
         }
     }
 
     @Override
     public CompletionStage<List<T>> invokeBatch(DataFetchingEnvironment dfe, Object[] arguments) {
 
-        Context vc = Vertx.currentContext();
-        if (runBlocking(dfe) || BlockingHelper.blockingShouldExecuteNonBlocking(operation, vc)) {
-            return super.invokeBatch(dfe, arguments);
-        } else {
-            return invokeBatchBlocking(dfe, arguments, vc);
+        ManagedContext requestContext = Arc.container().requestContext();
+        try {
+            RequestContextHelper.reactivate(requestContext, dfe);
+            Context vc = Vertx.currentContext();
+            if (runBlocking(dfe) || BlockingHelper.blockingShouldExecuteNonBlocking(operation, vc)) {
+                return super.invokeBatch(dfe, arguments);
+            } else {
+                return invokeBatchBlocking(dfe, arguments, vc);
+            }
+        } finally {
+            requestContext.deactivate();
         }
     }
 
@@ -83,23 +95,19 @@ public class QuarkusDefaultDataFetcher<K, T> extends DefaultDataFetcher<K, T> {
 
     @SuppressWarnings("unchecked")
     private CompletionStage<List<T>> invokeBatchBlocking(DataFetchingEnvironment dfe, Object[] arguments, Context vc) {
-        ManagedContext requestContext = Arc.container().requestContext();
-        try {
-            BlockingHelper.reactivate(requestContext, dfe);
-            SmallRyeThreadContext threadContext = Arc.container().select(SmallRyeThreadContext.class).get();
-            final Promise<List<T>> result = Promise.promise();
 
-            // We need some make sure that we call given the context
-            Callable<Object> contextualCallable = threadContext.contextualCallable(() -> {
-                return (List<T>) operationInvoker.invokePrivileged(arguments);
-            });
+        SmallRyeThreadContext threadContext = Arc.container().select(SmallRyeThreadContext.class).get();
+        final Promise<List<T>> result = Promise.promise();
 
-            // Here call blocking with context
-            BlockingHelper.runBlocking(vc, contextualCallable, result);
-            return result.future().toCompletionStage();
-        } finally {
-            requestContext.deactivate();
-        }
+        // We need some make sure that we call given the context
+        Callable<Object> contextualCallable = threadContext.contextualCallable(() -> {
+            return (List<T>) operationInvoker.invokePrivileged(arguments);
+        });
+
+        // Here call blocking with context
+        BlockingHelper.runBlocking(vc, contextualCallable, result);
+        return result.future().toCompletionStage();
+
     }
 
     private boolean runBlocking(DataFetchingEnvironment dfe) {

--- a/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/RequestContextHelper.java
+++ b/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/spi/datafetcher/RequestContextHelper.java
@@ -1,0 +1,22 @@
+package io.quarkus.smallrye.graphql.runtime.spi.datafetcher;
+
+import graphql.schema.DataFetchingEnvironment;
+import io.quarkus.arc.InjectableContext;
+import io.quarkus.arc.ManagedContext;
+
+public final class RequestContextHelper {
+
+    private RequestContextHelper() {
+    }
+
+    public static void reactivate(ManagedContext requestContext, DataFetchingEnvironment dfe) {
+        if (!requestContext.isActive()) {
+            Object maybeState = dfe.getGraphQlContext().getOrDefault("state", null);
+            if (maybeState != null) {
+                requestContext.activate((InjectableContext.ContextState) maybeState);
+            } else {
+                requestContext.activate();
+            }
+        }
+    }
+}

